lib/Reader/abisignature.cpp

```diff
@@ -580,7 +581,38 @@ Value *ABICallSignature::emitCall(GenIR &Reader, Value *Target, bool MayThrow,
   Type *FunctionTy = FunctionType::get(FuncResultType, ArgumentTypes, IsVarArg);
   Type *FunctionPtrTy = Reader.getUnmanagedPointerType(FunctionTy);
 
-  Target = Builder.CreateIntToPtr(Target, FunctionPtrTy);
+  // If we're passed a function value, we might need to update the type here
+  // to match, since we didn't know the right type when we created the function.
+  if (isa<llvm::Function>(Target)) {
+    llvm::Function *TargetFunc = cast<llvm::Function>(Target);
+    if (TargetFunc->getType() != FunctionPtrTy) {
+      bool MutateType = false;
+      // If we're going to Jmp to the target, don't mutate the type,
+      // since we may have added parameters to satisfy the constraints
+      // of LLVM's musttail that would cause this signature to diverge
+      // from what we'd have come up with in a normal call to the target.
+      if (!IsJmp) {
+        // We shouldn't change our view of function types multiple times,
+        // so only modify the type if it's the placeholder type we installed
+        // in makeDirectCallTargetNode.
+        Type *VoidType = Type::getVoidTy(Context);
+        PointerType *PlaceholderTy =
+            FunctionType::get(VoidType, false)->getPointerTo();
+        MutateType = TargetFunc->getType() == PlaceholderTy;
```

Comment:
I don't think it can happen. 

Basically if we see multiple possible types for a given method they all need to be compatible in some way -- same #s and kinds of arguments (modulo the weird jump case we exclude). Otherwise we're going to generate bad code no matter what we do.

`void()` is only compatible with itself.